### PR TITLE
Made tstruct a str -> Type mapping

### PR DESCRIPTION
**Pull request by cseed**
@konradjk this might break some of your stuff.

make tstruct a str -> Type mapping (and int -> Type mapping since tstructs are ordered).  tstruct.fields is still a tuple of fields, so to get the ith field, use t.fields[i].

### CODE REVIEW

**Comment by tpoterba** [comment]
already got conflicts!

File: python/hail/expr/functions.py [hunk 1]
```diff
@@ -633,14 +633,13 @@ def index(structs, identifier):
         raise TypeError("'index' expects an array with element type 'Struct', found '{}'"
                         .format(structs.dtype))
     struct_type = structs._type.element_type
-    struct_fields = {fd.name: fd.dtype for fd in struct_type.fields}
```
Comment (tpoterba) [inline python/hail/expr/functions.py]: we can just get rid of `index`. Is it used anywhere?

File: python/hail/expr/types.py [hunk 2]
```diff
@@ -716,51 +659,65 @@ def _convert_to_j(self, annotation):
...
-                for f in self.fields:
-                    f.dtype._typecheck(annotation.get(f.name))
+                                        (self, list(self), list(annotation)))
+                for f, t in self.items():
+                    t._typecheck(annotation.get(f))
             else:
                 raise TypeError("type 'struct' expected type Mapping (e.g. hail.genetics.Struct or dict), but found '%s'" %
                                 type(annotation))
 
+    def __getitem__(self, item):
```
Comment (tpoterba) [inline python/hail/expr/types.py]: add a typecheck_method(oneof(int, str))

File: python/hail/expr/functions.py (hunk 1 above)
Comment (cseed) [reply python/hail/expr/functions.py]: Ah, using the dict constructor.  niiice.

Comment (tpoterba) [reply python/hail/expr/functions.py]: I kinda remember it being used in the view joins. That might be a pain to try to do here.

Comment (cseed) [reply python/hail/expr/functions.py]: Yeah, I just discovered that.  I'm going to leave it in for now.

**Pull request merged**